### ISSUE TITLE
Replicated table data inconsistent between datanodes 

### ISSUE THREAD

**Issue by javanchueng** [postgres-x2/postgres-x2#256]
when i use postgres-x2-REL1_2_STABLE branch for testing ,
i found a problem (replication table data inconsistent between datanodes while pgxc cluster running under a certain high load and only using update statement) which can be reproduced by following steps :
1. 
   create table t1(id int not null,id1 int not null,name varchar(32),bt timestamp,score numeric(18,3)) distribute by replication;
   2.
   alter table t1 add constraint pk_t1 primary key(id,id1);
   3.
   insert into t1
   select a,b ,md5(random()::text),now(),10000
   from (select generate_series(1,100) as a) as aa,(select generate_series(1,10) as b) as bb;
2. vim test.sql
   \setrandom id 1 100
   \setrandom id1 1 10
   \setrandom score -5000 5000
   begin ;
   UPDATE t1 SET score = score + :score WHERE id = :id and id1=:id1 ;
   end ;
3. running pgbench with 3000 concurrent transactions
   nohup pgbench -c 1000 -T 10800 -n -j 1000 -r -f test.sql -h ipaddr -U postgres -p 5432 -d postgres  >> result1.txt &
   nohup pgbench -c 1000 -T 10800 -n -j 1000 -r -f test.sql -h ipaddr  -U postgres -p 5432 -d postgres  >> result2.txt &
   nohup pgbench -c 1000 -T 10800 -n -j 1000 -r -f test.sql -h ipaddr  -U postgres -p 5432 -d postgres  >> result3.txt &

while executing after 1.5hours errors occurs as below :
Client 569 aborted in state 4: ERROR:  duplicate key value violates unique constraint "pk_t1"
Client 273 aborted in state 4: ERROR:  duplicate key value violates unique constraint "pk_t1"
Client 418 aborted in state 4: ERROR:  Write to replicated table returneddifferent results from the Datanodes
Client 94 aborted in state 4: ERROR:  Write to replicated table returneddifferent results from the Datanodes
Client 915 aborted in state 4: ERROR:  Write to replicated table returneddifferent results from the Datanodes
client 765 executing \setrandomclient 441 executing \setrandom idClient 369 aborted in state 4: ERROR:  duplicate key value violates unique constraint "pk_t1"
Client 430 aborted in state 4: ERROR:  Write to replicated table returneddifferent results from the Datanodes
Client 258 aborted in state 4: ERROR:  Write to replicated table returneddifferent results from the Datanodes
Client 253 aborted in state 4: ERROR:  Write to replicated table returneddifferent results from the Datanodes

so is there anybody meet the same problem?

**Comment by Jasonysli** [comment]
We have met this problem before, but we can not reproduce it stablely. Thanks for your job!

**Comment by ghost** [comment]
We have meet the same problem. But  executing many hours maybe find the data is consistent, then executing many times, the result also is that the data is consistent. The problem is how to reproduce it stablely.

**Comment by koichi-szk** [comment]
Thanks for the info.   Let me find a time to run the same test.

Before this, could you let me know node configuration you need.   The
following info is helpful:
1. Number of coordinators/datanodes.
2. If you used gtm_proxy.
3. Is the environment local or not.

Thank you very much;

---

Koichi Suzuki

2015-07-01 10:26 GMT+09:00 javanchueng notifications@github.com:

**Comment by javanchueng** [comment]
sorry for the delay reply
i build 1 gtm master 3 gtm_proxy 3 coordinators 3 datanodes without any slaves in three servers
server1 : gtm_proxy1 coord1 db1
server2 : gtm_proxy2 coord2 db2
server3 : gtm_proxy3 coord3 db3 gtm_master

PS: i turned off auto_vacuum parameter

**Comment by javanchueng** [comment]
sorry for the delay reply
i build 1 gtm master 3 gtm_proxy 3 coordinators 3 datanodes without any slaves in three servers
server1 : gtm_proxy1 coord1 db1
server2 : gtm_proxy2 coord2 db2
server3 : gtm_proxy3 coord3 db3 gtm_master

PS: i turned off auto_vacuum parameter

**Comment by javanchueng** [comment]
sorry for the delay reply
i build 1 gtm master 3 gtm_proxy 3 coordinators 3 datanodes without any slaves in three servers
server1 : gtm_proxy1 coord1 db1
server2 : gtm_proxy2 coord2 db2
server3 : gtm_proxy3 coord3 db3 gtm_master

PS: i turned off auto_vacuum parameter

**Comment by javanchueng** [comment]
sorry for the delay reply
i build 1 gtm master 3 gtm_proxy 3 coordinators 3 datanodes without any slaves in three servers
server1 : gtm_proxy1 coord1 db1
server2 : gtm_proxy2 coord2 db2
server3 : gtm_proxy3 coord3 db3 gtm_master

PS: i turned off auto_vacuum parameter

**Comment by javanchueng** [comment]
sorry for the delay reply
i build 1 gtm master 3 gtm_proxy 3 coordinators 3 datanodes without any slaves in three servers
server1 : gtm_proxy1 coord1 db1
server2 : gtm_proxy2 coord2 db2
server3 : gtm_proxy3 coord3 db3 gtm_master

PS: i turned off auto_vacuum parameter

**Comment by javanchueng** [comment]
anybody have any update ?
not only replication table but also distribute by hash table have the same problem ;
i use benchmarksql-4.1.0 (4*512 concurrent transactions total 2048) test REL1_2_STABLE branch,
ERRORs:
" duplicate key value violates unique constraint pk_district " district table have duplicate key records;
district table have total 1000 rows ,after finishing the test ,995（maybe 996 not sure） rows left(benchmarksql only update the district table，no delete no insert sql statement)

i believe it caused by gxid inconsistency under high load ,but i have no idea how to fix it!

**Comment by koichi-szk** [comment]
Okay.   Sorry I didn't have a time to run the same test yet.

**Comment by koichi-szk** [comment]
XL people also encountered the same problem.   Pavan found a corner case of UPDATE handling in XC/XL, that is, there's some case where update serializability is broken, resulting in phantom.   We're discussing how to solve the issue.   It is not simple and will need new internal feature.

**Comment by jiangmj3** [comment]
Thanks for your reply, Could you give us some suggestion to discuss and fix
it, "there's some case where update serializability is broken" can you
explain deep detail for us. thanks a lot.

BRs
jim

2015-07-09 15:54 GMT+08:00 Koichi Suzuki notifications@github.com:

**Comment by koichi-szk** [comment]
Here's a quote from Pavan's analysis.

Phantom Records: While I could restore old behaviour and that fixed the first case, the phantom records issue seems like a different bug. From whatever analysis I've done so far, I see the following case. 

Say a tuple (version V) gets updated by transaction X1 (creating version V1) and the same in a very short while gets updated by another transaction X2 (creating version V2). For X2 to update the tuple, it must first see X1 as committed or aborted. Till then X2 will simply wait on X1.

Now for some reason, say snapshot issued by the GTM reports X1 as in-progress and X2 as committed, then the snapshot will see:
- version V as visible (since xmax X1 is still in-progress)
- version V1 as invisible or insert-in-progress ( xmin in-progress, xmax committed)
- version V2 as visible (xmin X2 is already committed, xmax invalid)

So two versions of the tuple will be returned.

Now the question why snapshot will show X1 as in-progress, but X2 as committed given that X1 must have committed first for X2 to be able to update version V1. I think I can see a corner case:

At coordinator:

Session 1: 
  Acquires X1
  Sends UPDATE command to the datanode
  (datanode finished updates, creating version V1)

  COMMIT X1
  PREPAREs X1 and reports PREPARED status to GTM
  COMMIT PREPARED on datanode
  (commit not yet reported to the GTM)       

Session 2:
  Acquires X2
  Sends UPDATE command to the detanode
  Datanode sees X1 committed, completes update, creating version V2
  COMMIT X2
  PREPARES X2 and reports PREPARED status to GTM
  COMMIT PREPARED on datanode
  Reports commit status to the GTM

Session 3:
  SELECT count(*) arrives
  Takes a snapshot from GTM
  GTM reports X1 as in-progress and X2 as committed in the snapshot

And then this snapshot can see both V and V2 as visible. I'm yet to fully establish this scenario, but thats what I am going to go after.

============= Koichi's addition to the above quote ===============================

We should block X2 above until X1 commit is reported to GTM and X1 does not appear to the snapshot.   I've proposed a couple idea to do this.

**Comment by jiangmj3** [comment]
thanks a lot, Do we have a fix plan? how can we  join it and help to do
something. if you have something need us to do ,please let me known.

BRs
jim

2015-07-10 11:28 GMT+08:00 Koichi Suzuki notifications@github.com:

**Comment by koichi-szk** [comment]
I think this is an issue to discuss in the next phone meeting, scheduled on July 14th.

**Comment by jiangmj3** [comment]
Hi Suzuki,  Can you share the result for this issue's discuss , thanks a
lot.

jim
BRs

2015-07-13 9:37 GMT+08:00 Koichi Suzuki notifications@github.com:

**Comment by koichi-szk** [comment]
I discussed the fix with Pavan and proposed three different alternatives.   One requires additional operations between coordinator and datanode, which could affect the performance.   The second needs some infrastructure for GTM, although there could be very small (not zero) overhead for normal operation.  Both looked expensive.

The third idea is to check if the transaction which updated the target record is still in the snapshot.   If so, it means that this (old) version is visible to other transactions, causing phantom.

So my idea is to wait until such gxid is not included in the snapshot freshly obtained from GTM.   So far, there's not further comment on this idea and I'm writing a code to do this.

To avoid any locks, such check is done in HeapTupleSatisfiesUpdate() function in tqual.c, the main code to check updating tuple visibility.

It's not very simple work but very feasible.

Hope this works well.   Anyway, I may ask for the test of this improvement when the code is done.

The following is very rough idea and design of the implementation with annotation I sent to Pavan and other XL people.   Any comments are welcome.   Please understand that the real code seems to be a bit different considering the module structure.

Regards;

# 

Attached is very rough preliminary design of the code.   Please allow typo, incorrect names and so on, unless they are significant.  This is just to present a design idea and does not compile successfully.

Also, please note that we need more careful review where GXID synchronization is needed in HeapTupleSatisfiesUpdate(), as indicated by
/\* Sync here */

Regards;

(Attached file)

```
#include “pgxc/xc_gtm_sync_timeout.h"

/* Enhance GetSnapshotGTM() and SetGlobalSnapshotData() to refresh this */
extrn GTM_Snapshot *latestGtmSnapshot;
/* XC needs dummy XID to get GTM_Snapshot.   XL doesn’t. */
static dummGxid = InvalidGlobalTransactionId;

bool
syncGXID_GTM(GlobalTransactionId gxid)
{
     int wait_tm = 0;

     if (latestGtmSnapshot == NULL)
          updateLatestSnapshot();
     if (!GTMSnapshotIncludes(latestGtmSnapshot, xid))
     {
          resetDummyGxid(true);
          return true;
     }
     while (wait_tm < xc_gtm_sync_timeout * 1000)     /* new GUC, in millisecond */
     {
          if (wait_tm)
               pg_usleep(wait_tm);
          if (wait_tm)
               wait_tm *= 2;
          else
               wait_tim = 10;     /* can be tuned */
          refreshLatestSnapshot();
          if (!GTMSnapshotIncludes(latestGtmSnapshot, xid))
          {
               resetDummyGxid(true);
               return true;
          }
     }
     resetDummyGxid(false);
     return false;
}

static void
refreshLatestSnapshot()
{
     if (dummyGxid == InvalidGlobalTransationId)
     {
          CheckConnection();
          dummyGxid = BeginTranGTM(timestamp);
     }
     (void)GetSnapshotGTM(dummyGxid, true);
}

static bool
GTMSnapshotIncludes(GTM_Snapshot snapshot, GlobalTransactinId gxid)
{
     /* これは正しい？gxmax を超える gxid が更新したデータってどうなる？これがコミットされているとうこと? */
     /* -> タイミングによってはあり得る。*/
     int i;
     GlobalTransactionId *xip;

      if (gxid < snapshot->gxmin || gxid > snapshot->gxmax)     /* Should use other functions to handle wraparound */
     /* If wraparound handling takes long, it may be better to skip this and just test sn_xip */
          return false;
     for (i = 0, xip = snapshot->sn_xip; i < snapshot->sn_cnt; xip++)
     {
          if (*xip == gxid)
               return true;
     }
     return false;
}

static void
resetDummyGxid(bool commit)
{
     if (dummyGxid == InvalidGxid)
          return;
     if (commit)
          CommitTranGTM(dummyGxid);
     else
           RollbackTranGTM(dummyGxid);
     dummyGxid = Invalidgxid;
     return;
}

— gtm.c ---

/*
 * If we migrate gtm_proxy to background worker, we can put this to the shared memory and
 * save some GTM interaction.  In this case, gtm_proxy can take care of this information update.
 */
GTM_Snapshot latestGtmSnapshot = NULL;

GTM_Snapshot
GetSnapshotGTM(GlobalTransactionId gxid, bool canbe_grouped)
{
    GTM_Snapshot ret_snapshot = NULL;
    CheckConnection();
    if (conn)
        ret_snapshot = get_snapshot(conn, gxid, canbe_grouped);
    if (ret_snapshot == NULL)
    {   
        latestGtmSnapshot = freeGTM_Snapshot(latestGtmSnapshot);
        CloseGTM();
        InitGTM();
    }   
     else
     {
          latestGtmSnapshot = setLatestGtm_Snapshot(latestGtmSnapshot, ret_snapshot);
     }
    return ret_snapshot;
}

static GTM_Snapshot
freeGTM_Snapshot(GTM_Snapshot snapshot)
{
     if (snapshot == NULL)
          return NULL;
     pfree(snapshot->sn_xip);
     pfree(snapshot);
     return NULL;
}

GTM_Snapshot
setLatestGtm_Snapshot(GTM_Snapshot target, GTM_Snapshot source)
{
     MemoryContext     old;
     GTM_Snapshot     r_snapshot;

     if (target == NULL)
          r_snapshot = (GTM_Snapshot)malloc(sizeof(GTM_Snapshotdata));
     else
     {
          r_snapshot = target;
          if (r_snapshot->sn_xip)
               free(r_snapshot->sn_xip);
     }
     memcpy(r_snapshot, source, sizeof(GTM_Snapshotdata));
     target->sn_xip = malloc(source->sn_xcnt * sizeof(GlobalTransactionData));
     memcpy(r_snapshot->sn_xip, source->sn_xip, source->xcnt*sizeof(GlobalTransactionData));
     return r_snapshot;
}

— procarray.c ---
void
SetGlobalSnapshotData(int xmin, int xmax, int xcnt, int *xip)
{
    snapshot_source = SNAPSHOT_COORDINATOR;
    gxmin = xmin;
    gxmax = xmax;
    gxcnt = xcnt;
    if (gxip)
        free(gxip);
    gxip = xip;
     /* Set latest global snapshot here */
     if (!latest_snapshot)
          lastest_snapshot = malloc(sizeof(GTM_Snapshot));
     else if (latest_snapshot->xip)
          free(latest_snapshot->dip);
     latest_snapshot->gxmin = gxmin;
     latest_snapshot->gmax = gxmax;
     latest_snapshot->sn_xcnt = gxcnt
     latest_snapshot->sn_xip = malloc(gxcnt * sizeof(GlobalTransactionId));
     memcpy(latest_snapshot->sn_xip, xip, latest_snapshot->sn_xcnt * sizeof(GlobalTransacctionId));

    elog (DEBUG1, "global snapshot info: gxmin: %d, gxmax: %d, gxcnt: %d", gxmin, gxmax, gxcnt);
}

 602 /*
 603  * HeapTupleSatisfiesUpdate
 604  *
 605  *  Same logic as HeapTupleSatisfiesNow, but returns a more detailed result
 606  *  code, since UPDATE needs to know more than "is it visible?".  Also,
 607  *  tuples of my own xact are tested against the passed CommandId not
 608  *  CurrentCommandId.
 609  *
 610  *  The possible return codes are:
 611  *
 612  *  HeapTupleInvisible: the tuple didn't exist at all when the scan started,
 613  *  e.g. it was created by a later CommandId.
 614  *
 615  *  HeapTupleMayBeUpdated: The tuple is valid and visible, so it may be
 616  *  updated.
 617  *
 618  *  HeapTupleSelfUpdated: The tuple was updated by the current transaction,
 619  *  after the current scan started.
 620  *
 621  *  HeapTupleUpdated: The tuple was updated by a committed transaction.
 622  *
 623  *  HeapTupleBeingUpdated: The tuple is being updated by an in-progress
 624  *  transaction other than the current transaction.  (Note: this includes
 625  *  the case where the tuple is share-locked by a MultiXact, even if the
 626  *  MultiXact includes the current transaction.  Callers that want to
 627  *  distinguish that case must test for it themselves.)
 628  */
 629 HTSU_Result
 630 HeapTupleSatisfiesUpdate(HeapTupleHeader tuple, CommandId curcid,
 631                         Buffer buffer)
 632 {
 633     if (!(tuple->t_infomask & HEAP_XMIN_COMMITTED))
 634     {
 635         if (tuple->t_infomask & HEAP_XMIN_INVALID)
 636             return HeapTupleInvisible;
 637
 638         /* Used by pre-9.0 binary upgrades */
 639         if (tuple->t_infomask & HEAP_MOVED_OFF)
 640         {
 641             TransactionId xvac = HeapTupleHeaderGetXvac(tuple);
 642
 643             if (TransactionIdIsCurrentTransactionId(xvac))
 644                 return HeapTupleInvisible;
 645             if (!TransactionIdIsInProgress(xvac))
 646             {
 647                 if (TransactionIdDidCommit(xvac))
 648                 {
                         /* Sync here */
 649                     SetHintBits(tuple, buffer, HEAP_XMIN_INVALID,
 650                                 InvalidTransactionId);
 651                     return HeapTupleInvisible;
 652                 }
                     /* Sync here */
 653                 SetHintBits(tuple, buffer, HEAP_XMIN_COMMITTED,
 654                             InvalidTransactionId);
 655             }
 656         }
 657         /* Used by pre-9.0 binary upgrades */
 658         else if (tuple->t_infomask & HEAP_MOVED_IN)
 659         {
 660             TransactionId xvac = HeapTupleHeaderGetXvac(tuple);
 661
 662             if (!TransactionIdIsCurrentTransactionId(xvac))
 663             {
 664                 if (TransactionIdIsInProgress(xvac))
 665                     return HeapTupleInvisible;
 666                 if (TransactionIdDidCommit(xvac))
                         /* Sync here */
 667                     SetHintBits(tuple, buffer, HEAP_XMIN_COMMITTED,
 668                                 InvalidTransactionId);
 669                 else
 670                 {
 671                     SetHintBits(tuple, buffer, HEAP_XMIN_INVALID,
 672                                 InvalidTransactionId);
 673                     return HeapTupleInvisible;
 674                 }
 675             }
 676         }
 677         else if (TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetXmin(tuple)))
 678         {
 679             if (HeapTupleHeaderGetCmin(tuple) >= curcid)
 680                 return HeapTupleInvisible;      /* inserted after scan started */
 681
 682             if (tuple->t_infomask & HEAP_XMAX_INVALID)  /* xid invalid */
 683                 return HeapTupleMayBeUpdated;
 684
 685             if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))    /* not deleter */
 686                 return HeapTupleMayBeUpdated;
 687
 688             if (tuple->t_infomask & HEAP_XMAX_IS_MULTI)
 689             {
 690                 TransactionId xmax;
 691
 692                 xmax = HeapTupleGetUpdateXid(tuple);
 693                 if (!TransactionIdIsValid(xmax))
 694                     return HeapTupleMayBeUpdated;
 695
 696                 /* updating subtransaction must have aborted */
 697                 if (!TransactionIdIsCurrentTransactionId(xmax))
                         /* Sync here */
 698                     return HeapTupleMayBeUpdated;
 699                 else
 700                 {
 701                     if (HeapTupleHeaderGetCmax(tuple) >= curcid)
 702                         return HeapTupleSelfUpdated;    /* updated after scan
 703                                                         * started */
 704                     else
 705                         return HeapTupleInvisible;      /* updated before scan
 706                                                         * started */
 707                 }
 708             }
 709
 710             if (!TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetRawXmax(tuple)))
 711             {
                     /* Sync here */
 712                 /* deleting subtransaction must have aborted */
 713                 SetHintBits(tuple, buffer, HEAP_XMAX_INVALID,
 714                             InvalidTransactionId);
 715                 return HeapTupleMayBeUpdated;
 716             }
 717
 718             if (HeapTupleHeaderGetCmax(tuple) >= curcid)
 719                 return HeapTupleSelfUpdated;    /* updated after scan started */
 720             else
 721                 return HeapTupleInvisible;      /* updated before scan started */
 722         }
 723         else if (TransactionIdIsInProgress(HeapTupleHeaderGetXmin(tuple)))
 724             return HeapTupleInvisible;
 725         else if (TransactionIdDidCommit(HeapTupleHeaderGetXmin(tuple)))
                 /* Sync here */
 726             SetHintBits(tuple, buffer, HEAP_XMIN_COMMITTED,
 727                         HeapTupleHeaderGetXmin(tuple));
 728         else
 729         {
 730             /* it must have aborted or crashed */
 731             SetHintBits(tuple, buffer, HEAP_XMIN_INVALID,
 732                         InvalidTransactionId);
 733             return HeapTupleInvisible;
 734         }
 735     }
 736
 737     /* by here, the inserting transaction has committed */
 738
 739     if (tuple->t_infomask & HEAP_XMAX_INVALID)  /* xid invalid or aborted */
             /* Sync here */
 740         return HeapTupleMayBeUpdated;
 741
 742     if (tuple->t_infomask & HEAP_XMAX_COMMITTED)
 743     {
 744         if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
                 /* Sync here */
 745             return HeapTupleMayBeUpdated;
 746         return HeapTupleUpdated;    /* updated by other */
 747     }
 748
 749     if (tuple->t_infomask & HEAP_XMAX_IS_MULTI)
 750     {
 751         TransactionId xmax;
   752
 753         if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
 754         {
 755             /*
 756             * If it's only locked but neither EXCL_LOCK nor KEYSHR_LOCK is
 757             * set, it cannot possibly be running.  Otherwise need to check.
 758             */
 759             if ((tuple->t_infomask & (HEAP_XMAX_EXCL_LOCK |
 760                                       HEAP_XMAX_KEYSHR_LOCK)) &&
 761                 MultiXactIdIsRunning(HeapTupleHeaderGetRawXmax(tuple)))
 762                 return HeapTupleBeingUpdated;
 763 
                 /* Sync here */
 764             SetHintBits(tuple, buffer, HEAP_XMAX_INVALID, InvalidTransactionId);
 765             return HeapTupleMayBeUpdated;
 766         }
 767
 768         xmax = HeapTupleGetUpdateXid(tuple);
 769         if (!TransactionIdIsValid(xmax))
 770         {
 771             if (MultiXactIdIsRunning(HeapTupleHeaderGetRawXmax(tuple)))
 772                 return HeapTupleBeingUpdated;
 773
 774             SetHintBits(tuple, buffer, HEAP_XMAX_INVALID, InvalidTransactionId);
                 /* Sync here */
 775             return HeapTupleMayBeUpdated;
 776         }
 777
 778         if (TransactionIdIsCurrentTransactionId(xmax))
 779         {
 780             if (HeapTupleHeaderGetCmax(tuple) >= curcid)
 781                 return HeapTupleSelfUpdated;    /* updated after scan started */
 782             else
 783                 return HeapTupleInvisible;      /* updated before scan started */
 784         }
 785
 786         if (MultiXactIdIsRunning(HeapTupleHeaderGetRawXmax(tuple)))
 787             return HeapTupleBeingUpdated;
 788
 789         if (TransactionIdDidCommit(xmax))
                 /* Wait here */
 790             return HeapTupleUpdated;
 791         /* it must have aborted or crashed */
             /* Sync here */
 792         SetHintBits(tuple, buffer, HEAP_XMAX_INVALID, InvalidTransactionId);
 793         return HeapTupleMayBeUpdated;
 794     }
 795
 796     if (TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetRawXmax(tuple)))
 797     {
 798         if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
 799             return HeapTupleMayBeUpdated;
 800         if (HeapTupleHeaderGetCmax(tuple) >= curcid)
 801             return HeapTupleSelfUpdated;        /* updated after scan started */
 802         else
 803             return HeapTupleInvisible;  /* updated before scan started */
 804     }
 805
 806     if (TransactionIdIsInProgress(HeapTupleHeaderGetRawXmax(tuple)))
 807         return HeapTupleBeingUpdated;
 808
 809     if (!TransactionIdDidCommit(HeapTupleHeaderGetRawXmax(tuple)))
 810     {
 811         /* it must have aborted or crashed */
 812         SetHintBits(tuple, buffer, HEAP_XMAX_INVALID,
 813                     InvalidTransactionId);
             /* Sync here */
 814         return HeapTupleMayBeUpdated;
 815     }
 816
 817     /* xmax transaction committed */
 818
 819     if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
 820     {
 821         SetHintBits(tuple, buffer, HEAP_XMAX_INVALID,
 822                     InvalidTransactionId);
 823         return HeapTupleMayBeUpdated;
 824     }
 825 
         /* Sync here */
 826     SetHintBits(tuple, buffer, HEAP_XMAX_COMMITTED,
 827                 HeapTupleHeaderGetRawXmax(tuple));
 828     return HeapTupleUpdated;    /* updated by other */
 829 }
```

#

**Comment by ghost** [comment]
Is same as the issue 239?

**Comment by koichi-szk** [comment]
Related but not the same.  #239 is the discussion when a transaction should report to GTM.   This issue is how to synchronize local commit with global commit.   As discussed in #239, we have to do the local commit first and then report to GTM.   This provides read consistency.

On the other hand, this issue is when such local commit should be visible to update.   According to Pavan's analysis and other experiments, local commit immediately allow update, even before local commit is reported to GTM and propagated to all the transactions.   This is serious corner case which I'm working at present.

Other observed inconsistency is issue #257, DDL inconsistency.   Yes, this #256 issue has some relationship to this.   Further, I suspect this can be caused by uncontrolled order of local operation.    I'm also working on this.   This fix, as seen in the issue track, is to use internal advisory lock to serialize DDL.

Obviously it is not an ideal fix.   Maybe we should predefine the order of local DDL operation as in the case of replicated table updates, but this needs bigger infrastructure and need some more resources.

Regards;
Koichi

**Comment by koichi-szk** [comment]
I changed snapshot refreshment schedule and found all the regression test passes.

I will continue regression test and may commit the patch if I'm confident it worth more intentional test.

The points so far are:
1. Most tuple reference (almost 99.9%) does not require any more test with fresh snapshot.
2. DDL requires longer period for synchronization.   Original default timeout value looks too short.
3. Original schedule between snapshot refreshment looks too short.

I will continue the test.   I'd like more people to join this issue and do any intentional test to see if there's more chance of data inconsistency.

I'm not sure if this fixes DDL inconsistency.   I have another idea to fix it, if there's any more DDL inconsistency observed with the patch.

Please give me some more days for my local test.

Regards;
Koichi

**Comment by jiangmj3** [comment]
Hi Suzuki，I‘m testing it now，when using benchmarksql to test and at the
same time to create table，there is no response

postgres=# select \* from pgxc_node;
 node_name | node_type | node_port | node_host  | nodeis_primary |
nodeis_preferred |   node_id
-----------+-----------+-----------+------------+----------------+------------------+-------------
 coord1    | C         |      5432 | localhost1 | f              | f
         |  1885696643
 coord2    | C         |      5432 | localhost2 | f              | f
         | -1197102633
 coord3    | C         |      5432 | localhost3 | f              | f
         |  1638403545
 coord4    | C         |      5432 | localhost6 | f              | f
         |  -890157977
 db1       | D         |     15432 | localhost1 | t              | t
         |    -2885965
 db2       | D         |     15432 | localhost2 | f              | f
         |   -79866771
 db3       | D         |     15432 | localhost3 | f              | f
         | -1472228704
 db4       | D         |     15432 | localhost6 | f              | f
         |  -972212898
(8 rows)

postgres=#

postgres=# select pg_backend_pid();

##  pg_backend_pid

```
      20202
```

(1 row)

postgres=# create table tt(id int);

gdb -p 20202

(gdb) bt
#0  0x00000031864eb197 in semop () from /lib64/libc.so.6
#1  0x00000000006f929c in PGSemaphoreLock (sema=0x7f07e85f4f50,
interruptOK=1 '\001') at pg_sema.c:421
#2  0x0000000000753800 in ProcSleep (locallock=0x2a4f5a0,
lockMethodTable=0x9f1960) at proc.c:1159
#3  0x000000000074e92c in WaitOnLock (locallock=0x2a4f5a0, owner=0x2b7a258)
at lock.c:1671
#4  0x000000000074dc16 in LockAcquireExtendedXC (locktag=0x7fffe6454710,
lockmode=5, sessionLock=0 '\000', dontWait=0 '\000', reportMemoryError=1
'\001', only_increment=0 '\000')
    at lock.c:1035
#5  0x000000000074d311 in LockAcquireExtended (locktag=0x7fffe6454710,
lockmode=5, sessionLock=0 '\000', dontWait=0 '\000', reportMemoryError=1
'\001') at lock.c:722
#6  0x000000000074d27a in LockAcquire (locktag=0x7fffe6454710, lockmode=5,
sessionLock=0 '\000', dontWait=0 '\000') at lock.c:681
#7  0x000000000074c6f4 in XactLockTableWait (xid=22007803) at lmgr.c:495
#8  0x00000000004ab60b in _bt_doinsert (rel=0x7f06cbf9fbf8, itup=0x2bd3728,
checkUnique=UNIQUE_CHECK_YES, heapRel=0x7f06cc017160) at nbtinsert.c:168
#9  0x00000000004b1a5a in btinsert (fcinfo=0x7fffe6454850) at nbtree.c:259
#10 0x0000000000889355 in FunctionCall6Coll (flinfo=0x2b781b0, collation=0,
arg1=139667168689144, arg2=140737056689328, arg3=140737056689296,
arg4=45954092, arg5=139667169177952, arg6=1)
    at fmgr.c:1440
#11 0x00000000004aa193 in index_insert (indexRelation=0x7f06cbf9fbf8,
values=0x7fffe6454cb0, isnull=0x7fffe6454c90 "", heap_t_ctid=0x2bd342c,
heapRelation=0x7f06cc017160,
    checkUnique=UNIQUE_CHECK_YES) at indexam.c:227
#12 0x0000000000508e9f in CatalogIndexInsert (indstate=0x2bd3138,
heapTuple=0x2bd3428) at indexing.c:136
#13 0x0000000000508f01 in CatalogUpdateIndexes (heapRel=0x7f06cc017160,
heapTuple=0x2bd3428) at indexing.c:162
#14 0x000000000052df12 in TypeCreate (newTypeOid=0, typeName=0x7fffe6455400
"tt", typeNamespace=2200, relationOid=16452, relationKind=114 'r',
ownerId=10, internalSize=-1, typeType=99 'c',
    typeCategory=67 'C', typePreferred=0 '\000', typDelim=44 ',',
inputProcedure=2290, outputProcedure=2291, receiveProcedure=2402,
sendProcedure=2403, typmodinProcedure=0,
    typmodoutProcedure=0, analyzeProcedure=0, elementType=0,
isImplicitArray=0 '\000', arrayType=16453, baseType=0,
defaultTypeValue=0x0, defaultTypeBin=0x0, passedByValue=0 '\000',
    alignment=100 'd', storage=120 'x', typeMod=-1, typNDims=0,
typeNotNull=0 '\000', typeCollation=0) at pg_type.c:450
#15 0x00000000005015c0 in AddNewRelationType (typeName=0x7fffe6455400 "tt",
typeNamespace=2200, new_rel_oid=16452, new_rel_kind=114 'r', ownerid=10,
new_row_type=0, new_array_type=16453)
    at heap.c:1317
#16 0x0000000000501a2e in heap_create_with_catalog (relname=0x7fffe6455400
"tt", relnamespace=2200, reltablespace=0, relid=16452, reltypeid=0,
reloftypeid=0, ownerid=10, tupdesc=0x2a53800,
    cooked_constraints=0x0, relkind=114 'r', relpersistence=112 'p',
shared_relation=0 '\000', mapped_relation=0 '\000', oidislocal=0 '\000',
oidinhcount=0, oncommit=ONCOMMIT_NOOP,
    reloptions=0, use_user_acl=1 '\001', allow_system_table_mods=0 '\000',
is_internal=0 '\000') at heap.c:1557
#17 0x00000000005ce656 in DefineRelation (stmt=0x2bd2a10, relkind=114 'r',
ownerId=10) at tablecmds.c:650
#18 0x000000000076d422 in ProcessUtilitySlow (parsetree=0x2b2c178,
queryString=0x2b2b550 "create table tt(id int);",
context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x2b2c548,
    sentToRemote=0 '\000', completionTag=0x7fffe6455c90 "") at
utility.c:1515
#19 0x000000000076cfd6 in standard_ProcessUtility (parsetree=0x2b2c178,
queryString=0x2b2b550 "create table tt(id int);",
context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x2b2c548,
    sentToRemote=0 '\000', completionTag=0x7fffe6455c90 "") at
utility.c:1361
#20 0x000000000076b1d6 in ProcessUtility (parsetree=0x2b2c178,
queryString=0x2b2b550 "create table tt(id int);",
context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x2b2c548,
    sentToRemote=0 '\000', completionTag=0x7fffe6455c90 "") at utility.c:361
#21 0x000000000076a2c1 in PortalRunUtility (portal=0x2bcfab0,
utilityStmt=0x2b2c178, isTopLevel=1 '\001', dest=0x2b2c548,
completionTag=0x7fffe6455c90 "") at pquery.c:1285
#22 0x000000000076a4ee in PortalRunMulti (portal=0x2bcfab0, isTopLevel=1
'\001', dest=0x2b2c548, altdest=0x2b2c548, completionTag=0x7fffe6455c90 "")
at pquery.c:1432
#23 0x00000000007699b5 in PortalRun (portal=0x2bcfab0,
count=9223372036854775807, isTopLevel=1 '\001', dest=0x2b2c548,
altdest=0x2b2c548, completionTag=0x7fffe6455c90 "") at pquery.c:882
---Type <return> to continue, or q <return> to quit---
#24 0x00000000007633a2 in exec_simple_query (query_string=0x2b2b550 "create
table tt(id int);") at postgres.c:1176
#25 0x0000000000767805 in PostgresMain (argc=1, argv=0x2a30420,
dbname=0x2a302d0 "postgres", username=0x2a302b0 "postgres") at
postgres.c:4323
#26 0x000000000070c165 in BackendRun (port=0x2a54c80) at postmaster.c:4264
#27 0x000000000070b8cb in BackendStartup (port=0x2a54c80) at
postmaster.c:3929
#28 0x00000000007081cc in ServerLoop () at postmaster.c:1729
#29 0x0000000000707a09 in PostmasterMain (argc=5, argv=0x2a2f540) at
postmaster.c:1390
#30 0x000000000066c0d1 in main (argc=5, argv=0x2a2f540) at main.c:227
(gdb)

2015-07-17 21:30 GMT+08:00 Koichi Suzuki notifications@github.com:

**Comment by koichi-szk** [comment]
Thank you jiangmingjun for the quick test.   The patch code is invoked only
from TransactionIdDidCommit() and TransactionIdIsKnownCompleted() so the
back trace itself does not show the problem.   Through the test, as I
reported, I found DDL may fail, but with the error that GXID
synchronization with GTM failed.    After additional test and improvement
of the patch, I found we can improve the synchronization schedule and
timeout value.

So in this case, because it seems to fail in semaphore synchronization,
there must be another process which cause this wait.

Could you do the following?
1.  In your reported situation, please test other processes, hopefully,
   other coordinators and datanodes which are taking care of CREATE TABLE
   statement.
2. Test the patch attached (this is the whole patch, so please remove the
   one you applied).  In using this patch, please specify the following line
   in all your postgresql.conf files.
   xc_gtm_commit_sync_test = on
   This will log how the patch works to all the log files.

The new patch has an improvement of default synchronization schedule and
timeout.

If you have the same situation, I'm afraid there could be semaphore
resource hang without any owner.   In this case, please try to clean the
semaphore explicitly.  You can do this by ipcs and ipcrm.

Thank you very much;

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-07-17 23:54 GMT+09:00 jiangmingjun notifications@github.com:

**Comment by koichi-szk** [comment]
Gee! I found attached file was removed!   Please take a look at
https://drive.google.com/open?id=0B88Ktv2MiEWGaVM4clZQU2pyMU0

Thanks!

**Comment by javanchueng** [comment]
hi Koichi Suzuki:
i did the test too! i use 2 benchmarksql client 1024 concurrent transactions ,after benchmarksql done successfully, the results shows as below:
TRACE   [2015-07-19 15:41:21.419]   Thread-49   Term-00, All terminals finished executing 2015-07-19 15:41:21
 INFO   [2015-07-19 15:41:21.419]   Thread-49   Term-00, 
 INFO   [2015-07-19 15:41:21.419]   Thread-49   Term-00, 
 INFO   [2015-07-19 15:41:21.419]   Thread-49   Term-00, Measured tpmC (NewOrders) = 1398.3
 INFO   [2015-07-19 15:41:21.419]   Thread-49   Term-00, Measured tpmTOTAL = 3104.43
 INFO   [2015-07-19 15:41:21.419]   Thread-49   Term-00, Session Start     = 2015-07-19 13:05:44
 INFO   [2015-07-19 15:41:21.419]   Thread-49   Term-00, Session End       = 2015-07-19 15:41:21
 INFO   [2015-07-19 15:41:21.420]   Thread-49   Term-00, Transaction Count = 483111
TRACE   [2015-07-19 15:41:21.420]   Thread-49   Term-00, Session finished!

TRACE   [2015-07-19 15:41:24.234]   Thread-294  Term-00, All terminals finished executing 2015-07-19 15:41:24
 INFO   [2015-07-19 15:41:24.234]   Thread-294  Term-00, 
 INFO   [2015-07-19 15:41:24.234]   Thread-294  Term-00, 
 INFO   [2015-07-19 15:41:24.234]   Thread-294  Term-00, Measured tpmC (NewOrders) = 2318.55
 INFO   [2015-07-19 15:41:24.234]   Thread-294  Term-00, Measured tpmTOTAL = 5143.69
 INFO   [2015-07-19 15:41:24.234]   Thread-294  Term-00, Session Start     = 2015-07-19 12:50:01
 INFO   [2015-07-19 15:41:24.234]   Thread-294  Term-00, Session End       = 2015-07-19 15:41:24
 INFO   [2015-07-19 15:41:24.234]   Thread-294  Term-00, Transaction Count = 881505
TRACE   [2015-07-19 15:41:24.235]   Thread-294  Term-00, Session finished!

but i can see the replication table data inconsistent between different nodes:
node 1：
postgres@localhost2:/home/postgres/pgxc_data/coord1/pg_log-->>psql -U benchmarksql -d postgres -p15432
psql (PGXC 3.0.7, based on PG 9.3.9)
Type "help" for help.

postgres=# select \* from district order by 1,2 limit 10;
 d_w_id | d_id |   d_ytd   | d_tax  | d_next_o_id |  d_name   |     d_street_1     |     d_street_2      |       d_city       | d_state |   d_zip  
--------+------+-----------+--------+-------------+-----------+--------------------+---------------------+--------------------+---------+-----------
      1 |    1 | 414449.04 | 0.1705 |        3280 | GsxMITO   | dlaKDGFUIbPUUsm    | oAfEKlWtGhqyfvnhZwB | kPkhuAAUabJveKxeez | WV      | 123456789
      1 |    2 | 442149.65 | 0.0537 |        3164 | FumwGCen  | VpLXWXdHSIyVCGlh   | gWiFOLVyKKyyj       | DDVEddqGKFynQZh    | NX      | 123456789
      1 |    3 | 359458.17 | 0.1659 |        3323 | lUrJnETis | vJsGxAdkyNXsKnrCB  | oNIfEnetIHO         | tAzHSDwQA          | KE      | 123456789
      1 |    4 | 397209.86 | 0.0408 |        3279 | qGqIdklV  | aCgvPGFJs          | kTVqOXaoKzIpGL      | BxIACGyoPzBeGR     | DZ      | 123456789
      1 |    5 | 401902.85 | 0.0904 |        3313 | yGHCYFLT  | xxuNgXsfytEyJIJi   | cdXHTcTxMieokoerPKP | FzPlmqTdTzf        | GH      | 123456789
      1 |    6 | 381280.88 | 0.0512 |        3311 | qFyGd     | PCugdQIwPVd        | YgxJridUKpmjwd      | pjgdQrvNLyJkXe     | JH      | 123456789
      1 |    7 | 373919.29 | 0.1314 |        3315 | dROPSkWE  | WkXsyVMPYavLmdp    | yXVJUzbKXrxHIdK     | MNxgIHwET          | SL      | 123456789
      1 |    8 | 417905.63 | 0.1666 |        3304 | hsuJL     | VPKOPnbEutIQkQnd   | tCErxFEOl           | ozsuQCqrhyZ        | HV      | 123456789
      1 |    9 | 420360.84 | 0.1794 |        3170 | zAttf     | qXtDrlhbP          | ZpAjVOdItquoROVd    | OWZPURczgcaTJGI    | PA      | 123456789
      1 |   10 | 313858.21 | 0.0193 |        3210 | iaJjY     | XdPMKLHDRSpWyQfUOG | YYyxCzTqFTloQnKQqk  | IdTySbgTxPLHvG     | OY      | 123456789
(10 rows)
postgres=# select count(1) from warehouse ;

##  count 

```
64
```

(1 row)

postgres=# select count(1) from district ;

##  count 

   995
(1 row)

postgres=# show xc_gtm_sync_timeout;

##  xc_gtm_sync_timeout 

 1ms
(1 row)

postgres=# show xc_gtm_commit_sync_test;

##  xc_gtm_commit_sync_test 

 on
(1 row)

node 2：
postgres@localhost3:/home/postgres/pgxc_data/db2/pg_log-->>psql -p15432 -U benchmarksql -d postgres 
psql (9.3.2, server 9.3.9)
Type "help" for help.

postgres=# 
postgres=# select \* from district order by 1,2 limit 10;
 d_w_id | d_id |   d_ytd   | d_tax  | d_next_o_id |  d_name   |     d_street_1     |     d_street_2      |       d_city       | d_state |   d_zip  
--------+------+-----------+--------+-------------+-----------+--------------------+---------------------+--------------------+---------+-----------
      1 |    1 | 414449.04 | 0.1705 |        3272 | GsxMITO   | dlaKDGFUIbPUUsm    | oAfEKlWtGhqyfvnhZwB | kPkhuAAUabJveKxeez | WV      | 123456789
      1 |    2 | 425054.99 | 0.0537 |        3164 | FumwGCen  | VpLXWXdHSIyVCGlh   | gWiFOLVyKKyyj       | DDVEddqGKFynQZh    | NX      | 123456789
      1 |    3 | 348020.82 | 0.1659 |        3316 | lUrJnETis | vJsGxAdkyNXsKnrCB  | oNIfEnetIHO         | tAzHSDwQA          | KE      | 123456789
      1 |    4 | 396552.76 | 0.0408 |        3276 | qGqIdklV  | aCgvPGFJs          | kTVqOXaoKzIpGL      | BxIACGyoPzBeGR     | DZ      | 123456789
      1 |    5 | 397829.27 | 0.0904 |        3304 | yGHCYFLT  | xxuNgXsfytEyJIJi   | cdXHTcTxMieokoerPKP | FzPlmqTdTzf        | GH      | 123456789
      1 |    6 | 372514.21 | 0.0512 |        3308 | qFyGd     | PCugdQIwPVd        | YgxJridUKpmjwd      | pjgdQrvNLyJkXe     | JH      | 123456789
      1 |    7 | 373313.73 | 0.1314 |        3311 | dROPSkWE  | WkXsyVMPYavLmdp    | yXVJUzbKXrxHIdK     | MNxgIHwET          | SL      | 123456789
      1 |    8 | 415000.37 | 0.1666 |        3297 | hsuJL     | VPKOPnbEutIQkQnd   | tCErxFEOl           | ozsuQCqrhyZ        | HV      | 123456789
      1 |    9 | 419018.31 | 0.1794 |        3170 | zAttf     | qXtDrlhbP          | ZpAjVOdItquoROVd    | OWZPURczgcaTJGI    | PA      | 123456789
      1 |   10 | 313858.21 | 0.0193 |        3210 | iaJjY     | XdPMKLHDRSpWyQfUOG | YYyxCzTqFTloQnKQqk  | IdTySbgTxPLHvG     | OY      | 123456789
(10 rows)

postgres=# select count(1) from warehouse ;

##  count 

   243
(1 row)

postgres=# select count(1) from district 
postgres-# ;

##  count 

  1674
(1 row)

postgres=# show xc_gtm_sync_timeout;

##  xc_gtm_sync_timeout 

 1ms
(1 row)

postgres=# show xc_gtm_commit_sync_test;

##  xc_gtm_commit_sync_test 

 on
(1 row)

At the same time i see error logs shown in my database :
2015-07-19 13:36:23.445 CST,"benchmarksql","postgres",22770,"10.1.226.201:43508",55ab301a.58f2,769,"UPDATE",2015-07-19 13:05:30 CST,400/480,2628832,ERROR,XX000,"Could not synchronize gxid (2627087) with GTM within 1 milliseconds.",,,,,,"UPDATE benchmarksql.warehouse SET w_ytd = w_ytd + $1  WHERE w_id = $2 ",,,""
2015-07-19 13:36:23.446 CST,"benchmarksql","postgres",23025,"10.1.226.201:43778",55ab3021.59f1,671,"UPDATE",2015-07-19 13:05:37 CST,535/375,2628529,ERROR,XX001,"Write to replicated table returneddifferent results from the Datanodes",,,,,,"UPDATE benchmarksql.warehouse SET w_ytd = w_ytd + $1  WHERE w_id = $2 ",,,""
2015-07-19 13:36:23.457 CST,"benchmarksql","postgres",28420,"10.1.226.201:42934",55ab2c6e.6f04,1626,"SELECT",2015-07-19 12:49:50 CST,137/1057,2628550,ERROR,XX000,"Could not synchronize gxid (2628543) with GTM within 1 milliseconds.",,,,,,"SELECT d_next_o_id, d_tax FROM benchmarksql.district WHERE d_id = $1 AND d_w_id = $2 FOR UPDATE",,,""

**Comment by koichi-szk** [comment]
It seems that 1ms timeout value is too short.    Anyway, the error message
indicates that the updating transactions aborted.

I wonder if such big difference occurs.   In regression, some tables are
replicated and updated (but not in concurrent manner) and they succeeds.

Do you see the similar difference without the patch?

---

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-07-19 17:47 GMT+09:00 javanchueng notifications@github.com:

**Comment by javanchueng** [comment]
um ,i have tried to set xc_gtm_sync_timeout =1ms 10ms 100ms 
the same results reproduced 
without the patch ,the same error "Write to replicated table returneddifferent results from the Datanodes"  reproduced 
The test scenario are as follows:
update the replication table very frequently and with high concurrent transactions;
we are doing marginal test ,althougth we know the limitation of using the replication table ;

**Comment by DBInsight** [comment]
hi Koichi Suzuki:

concurrent update is just one case, here is another example:

session1:
1.0 cn sent down insert statement to dn;
1.1 dn insert the tuple1 with xact xid1=10002, and in this time, the network between cn and gtm disconnected, 
     gtm detect network failure and remove xid1 from active transaction list, means xid1 is ended;
but dn is still working and executing xid1
 and network get right

session2:
2.0 get RencentXmin from gtm,in concurrent scenario, RecentXmin maybe larger than 10002, for example RecentXmin=10005, but this is wrong because it does not reflect the xid1 which is running in dn
2.1 session2 scan tuple which insert by tuple1
2.2 becuase of wrong RecentXmin, TransactionIdInProgress think xid1 is not running and see the clog, xid1 is not in committed status, so think it is rollbacked and mark tuple1 as XMIN_INVALID;

session1
1.2 xid1 is still running in dn, and dn come back to see tuple1, oh, my inserted tuple1 is not visible to me, so it maybe crash, or other problems

**Comment by DBInsight** [comment]
hi Koichi Suzuki:

i review your first patch,in TransactionIdDidCommit, the patch confirm if the xact commit on local, it must be commit on GTM, 

but what about prepared transaction, if one xact xid1 prepared on one node, and network wink, the gtm think it xid1 is ended, and the RecentXmin may be wrongly calculated, so it can also mark a prepared transaction as XMIN_INVALID, but later the prepared transaction maybe committed;

**Comment by koichi-szk** [comment]
Prepared transaction is not "COMMITTED" therefore TransactionIdDidCommit() is false.   The patch is applied only for transaction where TransactionIdDidCommit() is true.

Regards;

**Comment by DBInsight** [comment]
Maybe still has problems for prepared transaction whose status is transferring from "PREPARED" to "COMMITTED",  

 it also has a risk to mark a prepared transaction as HEAP_XMIN_INVALID(because it is not committed), actually later the  prepared transaction maybe committed

**Comment by koichi-szk** [comment]
As mentioned, these mechanism is the same as base PG.   GTM is
synchronizing these globally.   #256 is caused by the corner case of this
synchronization and the patch (although it need much more review and test)
is taking care of this.

---

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-07-29 12:17 GMT+09:00 DBInsight notifications@github.com:

**Comment by DBInsight** [comment]
not same as PG, we get snapshot from GTM, if GTM can not reflect all  active transactions which running on CN/DN in its snapshot, RecentXmin maybe wrong and cause TransactionIdInProgress get a wrong return value, and HeapSatisifyXXXX in tqual.c maybe  confuse

**Comment by koichi-szk** [comment]
Please understand the patch is filling this gap.

---

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-07-29 12:32 GMT+09:00 DBInsight notifications@github.com:

**Comment by DBInsight** [comment]
we should insist on one rule: ensure transaction status on gtm and on local node must be consistency;

so, there're several methods:
1st method: when we see TransactionIdIsCommitted on local node, we talk with gtm again to ensure it is also commit on gtm;
2nd method:inform gtm first, so when  TransactionIdIsCommitted is true on local node, it must be commit on gtm too; 

let me describe the 2nd method more detail:
session1(xid1):
1.1 cn sent prepare to nodes
1.2 nodes sent prepare ok to cn
1.3 cn inform gtm xid1 commit;
1.4 cn commit xid1 on local

if encounter any error(network wink or gtm restart cause gtm thread exit and remove xid1) in 1.3;
cn will rollback xid1 in 1.4;

if session2 comes in between 1.3-------1.4
2.1 session2 get a snapshot from gtm, 
2.2 session2 found that xid1 is not in this snapshot but running on cn, wait until xid1 commits done on local node;
2.3 when xid1 commits on local node, session2 continue

let me describe a risk for the 1st method:
session1(xid1):
1.1 cn sent prepare to nodes
1.2 nodes sent prepare ok to cn
1.2.1   **\* network wink or gtm restart ***
1.3 cn commit xid1 on local
1.4 cn inform gtm xid1 commit;

if there is a network wink between 1.2-----1.3, gtm detect the network failure and the corresponding thread on gtm exit and remove the xid1

session2 comes in between 1.2.1 --- 1.3
2.1 session2 get a snapshot from gtm which not contain xid1
2.2 seesion2 read tuple1 inserted by xid1, by HeapTupleSatisfyMVCC:
xid1 is not in progress and xid1 is not committed, so session2 consider xid1 aborted and mark tuple1 as HEAP_XMIN_INVALID,

after 1.3 cn commit xid1 on local, tuple1 still invisible because it has been marked HEAP_XMIN_INVALID

**Comment by koichi-szk** [comment]
I'll analyze the new input tomorrow.   Sorry I'm afraid I don't have a time today.

**Comment by DBInsight** [comment]
all querys are subject to gtm snapshot, when it get snapshot from gtm, if it found xid running on local node, but  committed on gtm, it will wait until xid1 commits done on local node;

**Comment by koichi-szk** [comment]
The patch worked well with Konstantin test and #266 test.  

Commits are as follows:
master: 2341c8c285fea7d8235424a4a357d93837b2b67a
REL1_2_STABLE: d79f5f06b54043d0aa18e7a276589d777584b9cb
REL1_1_STABLE: b85b75c7d2dda796d294559cb32a39f28d365653

DBInsight's additional comment is moved to new issue #274.

**Comment by koichi-szk** [comment]
Further observation revealed that the issue is not only in the heap tuple update.    The problem is related to the index check.

Original conversation in e-mail thread will be copied as the following comments.

**Comment by koichi-szk** [comment]
Here's a test script from jim guji586@gmail.com posted on Aug 6th, 2015.

create.sql:

```
create table t1(id int not null,id1 int not null,name varchar(32),bt timestamp,score numeric(18,3)) distribute by replication; 
alter table t1 add constraint pk_t1 primary key(id,id1); 
insert into t1 select a,b ,md5(random()::text),now(),10000 from (select generate_series(1,100) as a) as aa,(select generate_series(1,10) as b) as bb;
```

runtest.sh:

```
nohup ./test_t1.sh &
nohup ./test_t2.sh &
nohup ./test_t3.sh &
```

t1.sql:

```
¥setrandom id 1 100
¥setrandom id1 1 10
¥setrandom score -5000 5000
begin;
UPDATE t1 SET score = score + :score WHERE id = :id and id1=:id1 ;
end ;
```

t2:sql:

```
¥setrandom id 1 100
¥setrandom id1 1 10
¥setrandom score -5000 5000
begin ;
UPDATE t1 SET score = score + :score WHERE id = :id and id1=:id1 ;
end ;
```

t3.sql:

```
¥setrandom id 1 100
¥setrandom id1 1 10
¥setrandom score -5000 5000
begin;
UPDATE t1 SET score = score + :score WHERE id = :id and id1=:id1 ;
end ;
```

test_t1.sh:

```
pgbench -c 1000 -T 2400 -n -j 1000 -r -f t1.sql -h 10.1.238.1 -U javan -p 7432 -d postgres  >> result1.txt
```

test_t2.sh:

```
pgbench -c 1000 -T 2400 -n -j 1000 -r -f t2.sql -h 10.1.238.2 -U javan -p 7432 -d postgres  >> result2.txt
```

test_t3.sh:

```
pgbench -c 1000 -T 2400 -n -j 1000 -r -f t3.sql -h 10.1.238.3 -U javan -p 7432 -d postgres >> result3.txt
```

Instruction is as follows:

Hi Koichi

please see the attachment file.

```
    1. tar -xvf test_script.tar
    2  use crease.sql to create table and insert data
    3  nohup runtest.sh &
```

Jim's configuration:

```
postgres=# select * from pgxc_node;
 node_name | node_type | node_port | node_host  | nodeis_primary | nodeis_preferred |   node_id   
-----------+-----------+-----------+------------+----------------+------------------+-------------
 coord1    | C         |      7432 | localhost1 | f              | f                |  1885696643
 coord2    | C         |      7432 | localhost2 | f              | f                | -1197102633
 coord3    | C         |      7432 | localhost3 | f              | f                |  1638403545
 db1       | D         |     17432 | localhost1 | t              | t                |    -2885965
 db2       | D         |     17432 | localhost2 | f              | f                |   -79866771
 db3       | D         |     17432 | localhost3 | f              | f                | -1472228704
(6 rows)

postgres=# 
```

**Comment by koichi-szk** [comment]
Reply from Koichi on Aug 10, 2015.

Yes, I found the same message.   I observed a couple of rows are missing in one of the three datanodes.

It is not a phantom but missing record.

The previous patch takes eliminates phantom and the cause of this problem should be separate.  I need to look closer into other logs.   Interesting message is something like 

```
ERROR:  duplicate key value violates unique constraint "pk_t1"
DETAIL:  Key (id, id1)=(15, 1) already exists.
```

Because every statements are "UPDATE", and because no pkey update is found, such message should not be written.

I'll take a closer look to determine the exact cause.

Thank you for the report.

**Comment by koichi-szk** [comment]
From Jim, Aug 13th, 2015

Hi Koichi

Thanks a lot for your reply, Do you have some new found for the issue?
when we use the latest postgres-x2-REL1_2_STABLE to test replication table update , found another issue is that a lot of backend postgres process status is "UPDATE" and "UPDATE WAITING". and always can repetition.
if you want to found detail information, please use the "benchmarksql-4.1.0" to test it. 

BRs
jim

**Comment by koichi-szk** [comment]
From Koichi, Aug 19, 2015.

I added pgx2 developer mailing list to CC:

Yes, the issue happens in constraint violation checking.   It seems that the update begins before the former update has not been completely committed, I mean after local commit is done and before it is reported to GTM.   It is quite similar to the problem I fixed before, but this time it happens in BTREE updates.   I suppose similar phenomena will happen in other indexes, for example, hash, GIN and Gist.    We need to find a general way to synchronize the commit, not only in heap tuple but also in index records.

I don't think we need another benchmark to reproduce the same issue.   It is very helpful if you do some work to find similar issue in other indexes as I listed above.

Will work to find this specific and more general fix for index commit synch.

Thank you.

**Comment by koichi-szk** [comment]
From Jim, Aug 19, 2015

Hi koichi

very sorry to trouble you again.
Could you tell me the fix plan? thanks a lot
we use the patch to do a thorough test. found some issues.
the patch is only reduce the probability of a problem, not fundamentally solve the inconsistency of data.
so now we can not  use it in a business system . We can consider introducing distributed lock to keep consistency of data.

BRs
jim

**Comment by koichi-szk** [comment]
From Koichi, Aug 20, 2015.

Sorry it took long for the reply. 

The cause of the problem is very similar to "phantom" we saw in heap tuple, which has been fixed.

This phantom happened in the case of index tuple.   The cause is as follows:
1. Update to both heap and index tuple commits locally by the transaction T1, not not globally yet.
2. Second updating transaction (T2) comes and sees 1.
3. Because it is locally committed, T2 delete index tuple created by T1 and then T2 determines this is visible and checks unique constraint and finds it is a violation.
4. T2 aborts.
5. T1 is globally committed.  Because its xmax is set and valid, tuple updated by step1 above becomes invisible.

So the fix is to improve constraint violation check to include this global commit.

I don't think distributed lock works well.   This may kill the scalability.

I will work on this fix with high priority.

Thank you very much;

**Comment by koichi-szk** [comment]
From Jim, Aug 21, 2015.

thanks for your reply.
from your analyze information, I have some doubts.
when we update a tuple. the process flow like this:
heap_update -->ExecInsertIndexTuples.  first update data then update index
when heap update we will call syncGXID_GTM to sync local and global status.  when sync finish,  we can do index updated.
Otherwise we throw the error to abort the transaction.
I think it had included to check in global commit when constraint violation check.
if my idea is wrong, please let me know. thanks a lot. 

my second question is :
when use this patch to test, we found a lot of backend process status is "UPDATE WAITING"
whit gdb, we found it is blocked at PGSemaphoreLock 

```
(gdb) bt
#0  0x000000366b6eb197 in semop () from /lib64/libc.so.6
#1  0x00000000006f9478 in PGSemaphoreLock (sema=0x7fcc13a72ab0, interruptOK=1 '\001') at pg_sema.c:421
#2  0x00000000007539dc in ProcSleep (locallock=0x27a9690, lockMethodTable=0x9f1ca0) at proc.c:1159
#3  0x000000000074eb08 in WaitOnLock (locallock=0x27a9690, owner=0x28e3f90) at lock.c:1671
#4  0x000000000074ddf2 in LockAcquireExtendedXC (locktag=0x7fff4f855150, lockmode=5, sessionLock=0 '\000', dontWait=0 '\000', reportMemoryError=1 '\001', only_increment=0 '\000') at lock.c:1035
#5  0x000000000074d4ed in LockAcquireExtended (locktag=0x7fff4f855150, lockmode=5, sessionLock=0 '\000', dontWait=0 '\000', reportMemoryError=1 '\001') at lock.c:722
#6  0x000000000074d456 in LockAcquire (locktag=0x7fff4f855150, lockmode=5, sessionLock=0 '\000', dontWait=0 '\000') at lock.c:681
#7  0x000000000074c8d0 in XactLockTableWait (xid=78194) at lmgr.c:495
#8  0x000000000049a0d0 in heap_update (relation=0x7fbb91cf5168, otid=0x7fff4f8553c0, newtup=0x2913fb8, cid=1, crosscheck=0x0, wait=1 '\001', hufd=0x7fff4f855320, lockmode=0x7fff4f85531c) at heapam.c:3248
#9  0x000000000062d940 in ExecUpdate (tupleid=0x7fff4f8553c0, oldtuple=0x0, slot=0x2913418, planSlot=0x2912280, epqstate=0x290ed98, estate=0x290eb20, canSetTag=1 '\001') at nodeModifyTable.c:826
#10 0x000000000062e3ad in ExecModifyTable (node=0x290ecf0) at nodeModifyTable.c:1217
#11 0x000000000060c783 in ExecProcNode (node=0x290ecf0) at execProcnode.c:386
#12 0x000000000060a6b6 in ExecutePlan (estate=0x290eb20, planstate=0x290ecf0, operation=CMD_UPDATE, sendTuples=0 '\000', numberTuples=0, direction=ForwardScanDirection, dest=0xccfce0) at execMain.c:1497
#13 0x0000000000608a6a in standard_ExecutorRun (queryDesc=0x28d88b0, direction=ForwardScanDirection, count=0) at execMain.c:321
#14 0x000000000060895b in ExecutorRun (queryDesc=0x28d88b0, direction=ForwardScanDirection, count=0) at execMain.c:269
#15 0x0000000000768ca2 in ProcessQuery (plan=0x2894f40, sourceText=0x28d8790 "UPDATE benchmarksql.warehouse SET w_ytd = ((w_ytd)::double precision + $1) WHERE (w_id = $2)", params=0x28d8820, dest=0xccfce0, completionTag=0x7fff4f855860 "") at pquery.c:191
#16 0x000000000076a603 in PortalRunMulti (portal=0x27aa440, isTopLevel=1 '\001', dest=0xccfce0, altdest=0xccfce0, completionTag=0x7fff4f855860 "") at pquery.c:1386
#17 0x0000000000769b91 in PortalRun (portal=0x27aa440, count=9223372036854775807, isTopLevel=1 '\001', dest=0x27a3ef0, altdest=0x27a3ef0, completionTag=0x7fff4f855860 "") at pquery.c:882
#18 0x0000000000764cef in exec_execute_message (portal_name=0x27a3ae0 "", max_rows=9223372036854775807) at postgres.c:2103
#19 0x0000000000767bc9 in PostgresMain (argc=1, argv=0x278a4c8, dbname=0x278a3f0 "postgres", username=0x278a3d0 "benchmarksql") at postgres.c:4397
#20 0x000000000070c341 in BackendRun (port=0x27af030) at postmaster.c:4264
#21 0x000000000070baa7 in BackendStartup (port=0x27af030) at postmaster.c:3929
#22 0x00000000007083a8 in ServerLoop () at postmaster.c:1729
#23 0x0000000000707be5 in PostmasterMain (argc=5, argv=0x2789660) at postmaster.c:1390
#24 0x000000000066c2ad in main (argc=5, argv=0x2789660) at main.c:227
(gdb)   
```

**Comment by koichi-szk** [comment]
From Koichi, Aug 21, 2015.

Thanks for the update.    Are you quite sure that syncGXID_GTM() is called when heap is updated?   I'm afraid it might not be called.   This explains the issue.   Please let me know your thought.

If syncGXID_GTM() is called before heap is really updated, we need to look for other causes.

Sorry, I have to find a time to do more detailed analysis on this.

**Comment by koichi-szk** [comment]
From Jim, Aug 21, 2015.

```
(gdb) b heap_update
Breakpoint 3 at 0x499ba0: file heapam.c, line 2952.
(gdb) n
4305    if (ignore_till_sync && firstchar != EOF)
(gdb) n
4308    switch (firstchar)
(gdb) n
4580    TransactionId gxid = (TransactionId) pq_getmsgint(&input_message, 4);
(gdb) c
Continuing.

Breakpoint 1, ReadCommand (inBuf=0x7fff92e4fef0) at postgres.c:565
565 if (whereToSendOutput == DestRemote)
(gdb) c
Continuing.

Breakpoint 1, ReadCommand (inBuf=0x7fff92e4fef0) at postgres.c:565
565 if (whereToSendOutput == DestRemote)
(gdb) c
Continuing.

Breakpoint 1, ReadCommand (inBuf=0x7fff92e4fef0) at postgres.c:565
565 if (whereToSendOutput == DestRemote)
(gdb) c
Continuing.

Breakpoint 1, ReadCommand (inBuf=0x7fff92e4fef0) at postgres.c:565
565 if (whereToSendOutput == DestRemote)
(gdb) c
Continuing.

Breakpoint 2, syncGXID_GTM (gxid=721655147) at xact.c:5985
5985    int wait_tm = 0;
(gdb) bt
#0  syncGXID_GTM (gxid=721655147) at xact.c:5985
#1  0x00000000004cec14 in TransactionIdDidCommit (transactionId=721655147) at transam.c:167
#2  0x00000000008bb5bd in HeapTupleSatisfiesVacuum (tuple=0x7f6a2b1ffe18, OldestXmin=721655211, buffer=89) at tqual.c:1371
#3  0x00000000004a35e8 in heap_prune_chain (relation=0x7f6a28b51110, buffer=89, rootoffnum=1, OldestXmin=721655211, prstate=0x7fff92e4ea30) at pruneheap.c:372
#4  0x00000000004a320b in heap_page_prune (relation=0x7f6a28b51110, buffer=89, OldestXmin=721655211, report_stats=1 '\001', latestRemovedXid=0x7fff92e4f4ec) at pruneheap.c:198
#5  0x00000000004a3067 in heap_page_prune_opt (relation=0x7f6a28b51110, buffer=89, OldestXmin=721655211) at pruneheap.c:129
#6  0x00000000004aa892 in index_fetch_heap (scan=0x1a343f8) at indexam.c:527
#7  0x00000000004aa9cd in index_getnext (scan=0x1a343f8, direction=ForwardScanDirection) at indexam.c:616
#8  0x0000000000626ae1 in IndexNext (node=0x1a32fd8) at nodeIndexscan.c:78
#9  0x00000000006176a6 in ExecScanFetch (node=0x1a32fd8, accessMtd=0x6269a3 <IndexNext>, recheckMtd=0x626b03 <IndexRecheck>) at execScan.c:82
#10 0x0000000000617792 in ExecScan (node=0x1a32fd8, accessMtd=0x6269a3 <IndexNext>, recheckMtd=0x626b03 <IndexRecheck>) at execScan.c:167
#11 0x0000000000626baa in ExecIndexScan (node=0x1a32fd8) at nodeIndexscan.c:150
#12 0x000000000060cd4c in ExecProcNode (node=0x1a32fd8) at execProcnode.c:413
#13 0x000000000062e523 in ExecModifyTable (node=0x1a32cf0) at nodeModifyTable.c:1110
#14 0x000000000060cce3 in ExecProcNode (node=0x1a32cf0) at execProcnode.c:386
#15 0x000000000060ac16 in ExecutePlan (estate=0x1a32b20, planstate=0x1a32cf0, operation=CMD_UPDATE, sendTuples=0 '\000', numberTuples=0, direction=ForwardScanDirection, dest=0x1a2ead8) at execMain.c:1497
#16 0x0000000000608fca in standard_ExecutorRun (queryDesc=0x1a32710, direction=ForwardScanDirection, count=0) at execMain.c:321
#17 0x00007f6a389f0e00 in pgss_ExecutorRun (queryDesc=0x1a32710, direction=ForwardScanDirection, count=0) at pg_stat_statements.c:723
#18 0x0000000000608ea4 in ExecutorRun (queryDesc=0x1a32710, direction=ForwardScanDirection, count=0) at execMain.c:267
#19 0x00000000007692ca in ProcessQuery (plan=0x1a2bdc8, sourceText=0x18f4100 "UPDATE public.tt SET name = 'aaa'::character varying WHERE (id = 3)", params=0x0, dest=0x1a2ead8, completionTag=0x7fff92e4fce0 "") at pquery.c:191
#20 0x000000000076ac2b in PortalRunMulti (portal=0x18fa600, isTopLevel=1 '\001', dest=0x1a2ead8, altdest=0x1a2ead8, completionTag=0x7fff92e4fce0 "") at pquery.c:1386
#21 0x000000000076a1b9 in PortalRun (portal=0x18fa600, count=9223372036854775807, isTopLevel=1 '\001', dest=0x1a2ead8, altdest=0x1a2ead8, completionTag=0x7fff92e4fce0 "") at pquery.c:882
#22 0x0000000000763ba6 in exec_simple_query (query_string=0x18f4100 "UPDATE public.tt SET name = 'aaa'::character varying WHERE (id = 3)") at postgres.c:1176
#23 0x0000000000768009 in PostgresMain (argc=1, argv=0x18dab08, dbname=0x18da9a0 "postgres", username=0x18da980 "postgres") at postgres.c:4323
#24 0x000000000070c8b9 in BackendRun (port=0x18fa3c0) at postmaster.c:4264
#25 0x000000000070c01f in BackendStartup (port=0x18fa3c0) at postmaster.c:3929
#26 0x0000000000708920 in ServerLoop () at postmaster.c:1729
#27 0x000000000070815d in PostmasterMain (argc=4, argv=0x18d9bc0) at postmaster.c:1390
#28 0x000000000066c825 in main (argc=4, argv=0x18d9bc0) at main.c:227
(gdb) n
5986    int wait_total = 0;
(gdb) n
5987    int retry = 0;
(gdb) n
5989    if (!shouldSyncGXID)
(gdb) p shouldSyncGXID
$2 = 1 '\001'
(gdb) l
5984    {
5985    int wait_tm = 0;
5986    int wait_total = 0;
5987    int retry = 0;
5988
5989    if (!shouldSyncGXID)
5990    return false;
5991
5992    if (latestGTMSnapshot == NULL)
5993    {
(gdb) l
5994    if (xc_gtm_commit_sync_test)
5995    elog(LOG, "GXID(%u): No latestGTMSnapshot, refreshing.", gxid);
5996    refreshLatestGTMSnapshot();
5997    retry++;
5998    }
5999    if (!gtmSnapshotIncludes(latestGTMSnapshot, gxid))
6000    {
6001    if (xc_gtm_commit_sync_test)
6002    elog(LOG, "Sync GXID(%u): refresh(%d), No further refreshLatestGTMSnapshot() needed.", gxid, retry);
6003    resetDummyGxid(true);
(gdb) n
5992    if (latestGTMSnapshot == NULL)
(gdb) n
5999    if (!gtmSnapshotIncludes(latestGTMSnapshot, gxid))
(gdb)
```

**Comment by koichi-szk** [comment]
From Koichi, Aug 25, 2015.

I'm analyzing ExecUpdate() function.

Yes, as you pointed out, syncGXID_GTM() has been called when updating the heap.   Unfortunately, it has nothing to do with the index used to check the uniqueness.

So even though the heap is updated correctly, there still could be a chance that the index remains unchanged and violation within the index can be observed.    In the current code it could happen when HOT does not work and updated tuple is written to another page.

I'm analyzing how it happens and why it doesn't happen in vanilla PG.

Also I found this is common to all the indexes and we may be able to find a fix applicable to all the indexes.

Because it is different cause, I'll move this topic to new issue and copy all the posted materials to there.

Sorry for taking long.

Thank you.

**Comment by koichi-szk** [comment]
I returned to `_bt_check_unique()` and found that violation check uses the same snapshot as original Heap Tuple visibility check.   We have a good chance that some tuple, which is not visible at heap tuple level with `syncGXID_GTM()` is regarded still visible here.   In this case, the tuple is determined as "constraint violation" and the updating transaction aborts.   On the other hand, another tuple, determined as conflicting, is eventually deleted and is gone.

May be this is the cause.

The fix can be btree specific.   At present, unique constraint is implemented only with btree so we may not need any more fix.

I will continue the work to post the fix.

**Comment by koichi-szk** [comment]
Update of the analysis.   So far, the issue can be caused as follows:

Update takes place as follows:

a) Delete old tuple (update xmax of the old tuple)
b) Insert new tuple
c) Check if old tuple is dirty, that is, xmax of the old tuple is included in the snapshot.

Considering the same cause as phantom in replicated table updates without unique constraint, there could be a chance for the reported situation to happen, especially in checking BTree uniquenes.

It is not straightforward and need some more analysis of the details of the code in _bt_check_unique and associated snapshot checking.

Sorry taking this long.   Hope to locate the cause exactly shortly.

**Comment by koichi-szk** [comment]
To fix above, I mean, to tell _bt_check_constraint() correctly, I suppose we should remove synchronized gxid from the original snapshot.    Unfortunately, it is not simple to do this because commit synchronization does not come with the snapshot used, just gxid value.   So to feed back this synchronization, we need something more.   Current transaction is the most one used but it is copied for other usage so I need to determine what snapshot to maintain.

May take a bit more.   It is not simple.

**Comment by koichi-szk** [comment]
I made another addition to syncGXID_GTM().    Because this function only waits until the specified GXID commit/abort has been reported to GTM.   This GXID remains in the current snapshot and I thought this can cause another inconsistency in the visibility at constraint check.

Additional code is to remove the gxid from current/secondary snapshot.

This does not work so far.   Will take more information to see what is the real cause.

Sorry taking so long.

**Comment by koichi-szk** [comment]
I did the same test with the current head of REL1_2_STABLE.   Yes, we had duplicate key value error but such dirty transactions are aborted and then the resulting database looks consistent.

Please see the result:

```
koichi=# execute direct on (datanode1) 'select count(*) from t1';
 count 
-------
  1000
(1 row)

koichi=# execute direct on (datanode2) 'select count(*) from t1';
 count 
-------
  1000
(1 row)

koichi=# execute direct on (datanode3) 'select count(*) from t1';
 count 
-------
  1000
(1 row)

koichi=# 
```

Hope someone does the same test and share the result.

I'll do the same test with the master.   Because the master does not contain any further fixes done to PG branches, there could be PG-originated bugs too.

Regards;

**Comment by koichi-szk** [comment]
With the master (although I some node stopped somehow), I had the same result as above.

I suspect that the conflict could occur in very busy status but such conflict can be resolved by aborting one of the conflicting transactions (as done in deadlocks).

It's nice if somebody run the same test.   Duplicate key could be okay.   Issue is if the whole database can be maintained consistent.

Regards;

**Comment by koichi-szk** [comment]
I have been waiting for three weeks for more input on this test.   As the patch looks working and no more failure reports have been posted, I'll close this.
